### ISSUE TITLE
DataTable: when it is updated, the filter values are set empty. (LazyDataModel + "special" API-usage)

### ISSUE THREAD

**Issue by mertsincan** [primefaces/primefaces#9349]
As in the example below, after selecting a value from the selectcheckbox menu, the filter values are lost when the table is updated. It works fine with version 11.
```
<h:form>
      <p:dataTable var="customer" value="#{dtLazyView.lazyModel}" paginator="true" rows="10"
                   paginatorTemplate="{RowsPerPageDropdown} {FirstPageLink} {PreviousPageLink} {CurrentPageReport} {NextPageLink} {LastPageLink}"
                   rowsPerPageTemplate="5,10,15" selectionMode="single"
                   selection="#{dtLazyView.selectedCustomer}"
                   id="customerTable" widgetVar="_customerTable">

        <p:column headerText="Name" sortBy="#{customer.name}">
          <h:outputText value="#{customer.name}"/>
        </p:column>
        <!-- This filter hides correctly, but is not working when updated -->
        <p:column headerText="country " sortBy="#{customer.country}" filterBy="#{customer.country.name}">
          <f:facet name="filter">
            <p:selectCheckboxMenu id="multiple" value="#{dtFilterView.selectCountries}" style="min-width: 15rem"
                                  multiple="true" scrollHeight="250" updateLabel="true">
              <p:ajax event="change" process="@this" update="@form:customerTable"/>
              <f:selectItems value="#{dtLazyView.countries}"/>
            </p:selectCheckboxMenu>
          </f:facet>
          <h:outputText value="#{customer.country.name}"/>
        </p:column>
      </p:dataTable>

    </h:form>
```

**Comment by tandraschko** [comment]
@mertsincan that reverts my fix for: https://github.com/primefaces/primefaces/issues/9040

**Comment by mertsincan** [comment]
Hi @tandraschko, yes but unfortunately, there are at least 5 issues reported by PROs regarding https://github.com/primefaces/primefaces/issues/9040 
When they update the PF version, they completely lose the value they submitted in the filter. 
Exp; 
[primefaces-test.zip](https://github.com/primefaces/primefaces/files/9879004/primefaces-test.zip)
Maybe, you can have a look at this example to be compatible with your fix.
Note: As I said above, it works fine with PF11, PF10 and 8.

**Comment by christophs78** [comment]
We need to add integrationtests for this specific situation. (or extend exisiting ones)

**Comment by christophs78** [comment]
Needed some time to at least understand to some degree what´s going on there while i tried to add a good integrationtest. (Don´t ask what i did yesterday evening...)

The reproducer provided by the PROs is at least somehow awkward...

```java
@Override
public List<Customer> load(int offset, int pageSize, Map<String, SortMeta> sortBy, Map<String, FilterMeta> filterBy) {
  // apply offset & filters
  List<Customer> customers = datasource.stream()
      .skip(offset)
      .filter(o -> filterView.getSelectCountries().isEmpty() || filterView.getSelectCountries().stream().anyMatch((country) -> country.equals(o.getCountry().getName())))
      .limit(pageSize)
      .collect(Collectors.toList());

  return customers;
}
```

This code ignores filterBy provided to `LazyDataModel#load`. Instead this code references to a ViewScoped view-bean. Maybe it work´s, but the customer does not use the API the way it´s intended.

**Comment by djmj** [comment]
> This code ignores filterBy provided to `LazyDataModel#load`. Instead this code references to a ViewScoped view-bean. Maybe it work´s, but the customer does not use the API the way it´s intended.

I see the problem now and had some thoughts.

**Use case**

We wanted to use our own "filter" with value binding and updating the table with ajax. In our load method we use the value, similar to your usecase.

Of cause we want our "filter" to be displayed in the table header. This does not work so good by just using the _header_ facet since it displays the sort indicator below it.

So we revert to the _filter_ facet but its not showing up because we have to speficy _filterBy_ attribute.

**Idea?**

Allow displaying of facet _filter_ even if no _filterBy_ attribute is present. (I dont know if this has any side effects?)
In this case developer is responsible of providing filtering himself.

This still does not solves the issue that even if someone wants to use _filterBy_ as intended and  intentional still wants to ajax bind this value (to reuse it for something else), the value will be lost.

**Comment by Rapster** [comment]
I can try to give my thoughts on this. First, I don't see any good reason to have a null check on the filterValue, knowing the String madness in a JSF environnement, we should allow to take null or empty string as an input. Second, I can easily imagine a scenario where I decide to reset the filter value using  the "filter api" even though I just set one (edit: turns out, looks like to be #9040 scenario 😅 )... So on very theoretical note, I think we should revert that fix.

Now, based on the reproducer (and only, I'll try to come back on @djmj coment), Filter feature is not used in that use case (at least, not the one provided by PF). Since countries can be supported by PF built-in feature, why not using it? It'll work and less code :wink: Maybe we need some context here, what exactly your customer wants  to do...

@djmj I'm still confused about what exactly you're trying to achieve. Do you want to display your filters somewhere else than columns? Maybe a reproducer would be very helpful to solve that and your issue

But as I far as I can see, since there is no clear use case, I would advise to revert that commit and wait till clear explanation is provided. Remember to always git blame or have a quick look at the git history to see if you're not reverting anything

**Comment by djmj** [comment]
@Rapster

Here some more details what we try to achieve.

**Example 1 - reuse value in load method**

Sometimes we have certain "filter" values binded to bean properties because we have to reuse them inside the load method. Below a simplified example where the binded value is passed to the service remote method.

**Example 2 - reuse value to preselect other options**

Further as an example the value is passed to a new instance of `Car` which can be created and persisted in database in another form. The year is already preselected for the user this way. 

This is not uncommon in master detail pages. Example with computer hardware is very restrictive. 
If you select CPU AMD you need a motherboard of AMD ... so you help the user preselect the filter by value binding.

The user saves time this way.

**LazyDataModel**

	class CarModel extends LazyDataModel<Car>
	{
		private Integer year;
		private RemoteBean remoteBean;

		private Car createCar = new Car();

		@Override
		public List<Car> load(final int first, final int pageSize, final Map<String, SortMeta> sortMetas,
			final Map<String, FilterMeta> filters)
		{
			return remoteBean.loadCars(first, pageSize, year);
		}

		// getter 

		public void setYear(Integer year)
		{
			this.year = year;

			// preselect year for a new car
			this.createCar.setYear(year);

			// or pass year to another LazyDataModel to "preselect" a filter
		}
	}

**Markup**

	<p:column headerText="Year" filterBy="#{varCar.year}">
		<f:facet name="filter">
			<p:inputText id="year" value="#{bean.carModel.year}" converter="javax.faces.Integer">
				
				<p:ajax update="@form"/>
			</p:inputText>
		</f:facet>
	</p:column>

**Why are we not using the provided _filterMeta_?**

Because the service method needs an integer `year` and writing something like `Integer year = (Integer)filterMeta.get("year")` with casting is no nice code. This example is simplified and sometimes you have 10 params to get from the map like this since you cannot change the service method you call.  We also mix it with filters outside the table if they have no column representation. 

For our standard cases we use `FilterMeta` but never look inside it. We have our own framework which converts `FilterMeta` to serializable query data and passes it to the remote service. This service converts this query data to Hibernate Criteria API / HQL and is completely generic. Similar to fairly new JPA lazy data model but serializable and works with remote services. 

**Why are we using _filterBy_ if we dont use `year` from _filterMeta_?**

Cause we dont have any good options to place our _own_ filters in the column header. Facet `header` is not suitable for it.

**Comment by Rapster** [comment]
What about this then?

```xml
<p:column headerText="Year" filterBy="#{varCar.year}">
	<f:facet name="filter">
		<p:inputText id="year" value="#{bean.carModel.year}" converter="javax.faces.Integer">
			<!-- process year to bean first and then update table using PF API -->
			<p:ajax oncomplete="PF('table').filter'()" />
		</p:inputText>
	</f:facet>
</p:column>
```

You just need to process first and then update, instead of `onchange` (which would be called first) use `oncomplete` event (to ensure that your bean has been processed first).

**Comment by djmj** [comment]
I see and this works when user is changing the filter but not when the filter has a default value in the bean set. For example we set the year filter by default to 2022. 

	private Integer year = 2022;

At page load its not set and any other ajax request updating the table will not set it.

**Comment by Rapster** [comment]
What about `FilterMeta#filterValue` then?

> writing something like Integer year = (Integer)filterMeta.get("year") with casting is no nice code.

I don't see the "problem" here, looks fine by me...

**Comment by djmj** [comment]
Looks like old PHP map fizzling to me and its not DRY. I have to repeat the String literals from the markup `"year"` inside the java code. I try to avoid this.

We dont need a property `year` since we already have an existing filter bean like `CarFilter` with many properties provided by a shared library:

	private CarFilter carFilter;
		
	// load method - unnecessary boilerplate filter "set" calls
	public List<T> load(...)
	{
		carFilter.setYear((Integer)filterMeta.get("year").getFilterValue());
		carFilter.setMonth((Integer)filterMeta.get("month").getFilterValue());
		carFilter.setFooBarName((String)filterMeta.get("foo.bar.name").getFilterValue());

		return remoteBean.load(first, pageSize, carFilter);
	}
This filter instance is passed to the remote bean so its straightforward for us to reuse it with value binding. 

*Generic*

To take this further the model could be completely generic with value bindings.

	class FilterModel<T, V> extends LazyDataModel<T>
	{
		private V filter;

		// load method
		return remoteBean.load(first, pageSize, filter);
	}

The markup just handles everything and in the Bean its just a one liner.




The issue is not style of code or taste but why "shouldnt" filter be able to bind its value? Its an input component overall like any other `<p:inputText>`.

**Comment by Rapster** [comment]
Sure... Please come up with a reproducer, it'll save all of us a bunch of time because so far I don't see why PF can't handle your use case

**Comment by christophs78** [comment]
Even better - provide a (probably failing) integration test. Use https://github.com/primefaces/primefaces/commit/e7333816c420b260c504f8bc9238ae7059902521 as kind of template.
(This way we ensure not to break existing usecases. PrimeFaces, JSF and DataTable allow for so many usecases. Integration tests help us to avoid breaking other existing usecases while trying to fix or enhance other ones.)

**Comment by melloware** [comment]
+1 to @Rapster @christophs78   suggestions.  @djmj even if you put together a pF Test showing the issue we can always turn it into an Integration Test.  We just need to see your actual use case.

**Comment by djmj** [comment]
The reproducer in https://github.com/primefaces/primefaces/issues/9349#issuecomment-1293407918 covers the same use case we have. Only difference is we are calling a remote service instead doing filtering in memory. I gave arguments why we are using filters with value binding and not the API intended way.

**Comment by Rapster** [comment]
Ok, so based on the reproducer that Mert provided. I understand two issues:
- once one/several value(s) from the custom filter is/are selected, it/they gets lost in column filter
- if you have preselected values in the custom filter (e.g selectCheckboxMenu here), these are not showing in the filter therefore not taken account in load method at page loading

Do I get that right?

**Comment by djmj** [comment]
Yes that is correct. 

**Simplified example**

Here simple test files i send @mertsincan at PRO support. In this test intentional i show the `filterYear` property again in the `header` facet and also use a second component outside the table to double check.

This shows, that the `filterYear` is correctly pushed to the bean but on update the input component in the filter facet is not updated with that value.

	<p:column filterBy="#{car.year}" sortBy="#{car.year}">
		<f:facet name="header">
			
			<h:outputText value="Year: #{testDataTableFilterAjaxBean.filterYear}" />
		</f:facet>
					
		<f:facet name="filter">
									
			<p:selectOneMenu id="filterYear" value="#{testDataTableFilterAjaxBean.filterYear}" size="4" > 
				<f:selectItem itemLabel="None" itemValue="#{null}" />
				<f:selectItems value="#{testDataTableFilterAjaxBean.years}" />		
				<p:ajax process="@this" update="@form"/>							
			</p:selectOneMenu>
		</f:facet>
					
		<h:outputText value="#{car.year}" />
	</p:column>

<img width="756" alt="test-datatable-onemenu-ajax-lostvalue-client" src="https://user-images.githubusercontent.com/5265550/200079759-dfc22f71-da42-443f-a7c4-16a7caa9176d.png">


[test-dataTableFilter-ajax-lost.zip](https://github.com/primefaces/primefaces/files/9942154/test-dataTableFilter-ajax-lost.zip)

**Comment by Rapster** [comment]
Ok, so again based on Mert reproducer (sorry, can't pull altogether the files you submitted, that's why we ask for reproducer). Solutions has been handed over to you [here](https://github.com/primefaces/primefaces/issues/9349#issuecomment-1299789421) and [here ](https://github.com/primefaces/primefaces/issues/9349#issuecomment-1301103383) _(my bad I mentionned FilterMeta when it should be UIColumn, but it's kinda the same thing)_. As a result, that's what it looks like:

```xml
<p:column headerText="country " sortBy="#{customer.country}" filterBy="#{customer.country.name}" filterValue="#{dtLazyView.lazyModel.countries}">
    <f:facet name="filter">
        <p:selectCheckboxMenu id="multiple" value="#{dtLazyView.lazyModel.countries}" style="min-width: 15rem"
                                  multiple="true" scrollHeight="250" updateLabel="true">
            <p:ajax process="@this" oncomplete="PF('_customerTable').filter()" />
            <f:selectItems value="#{dtLazyView.countries}"/>
        </p:selectCheckboxMenu>
    </f:facet>
    <h:outputText value="#{customer.country.name}"/>
</p:column>
```

You only have to tweak the filterValue so when it gets cleared, you returned the whole countries collection

[primefaces-test.zip](https://github.com/primefaces/primefaces/files/9942283/primefaces-test.zip)

**Comment by Rapster** [comment]
Since there are no feedbacks on this and I don't see any issue with 12.0  and the provided scenario, maybe the commit can be reverted WDYT?

**Comment by melloware** [comment]
I am all for reverting it but its up to @mertsincan ??

**Comment by tandraschko** [comment]
+1 for revert it

if we really required, we can also add a attribute to DT

**Comment by christophs78** [comment]
Reverting will break DataTable039Test (https://github.com/primefaces/primefaces/commit/e7333816c420b260c504f8bc9238ae7059902521). We at least need to modify (or disable) this as part of the revert-PR.

**Comment by Rapster** [comment]
And if you apply the fix proposed https://github.com/primefaces/primefaces/issues/9349#issuecomment-1304304966 Does it work?

**Comment by christophs78** [comment]
We before added https://github.com/primefaces/primefaces/commit/8b40e57af079e86323ac5cc98eb35fd2977e5ad8#diff-5f99e9fbde103f8ae481c72952b9c72c402b4471d178ddaa5208f7ed2508ef17. This one is similar but not exact the same.
You told us, your suggested solution based on PF-test works. So it should also work to put this into a integrationtest.
(Overall the PF history has IMO too many instances of fixing one thing and breaking another one. We should get out of this cycle.)

**Comment by mertsincan** [comment]
Sorry, I missed these comments :/ I still don't understand that users have to use an attribute like filterValue. I think the way of use in the project I attached above is the kind of use we usually encounter in every project. If you are updating the table in the filter, I think it is not the right approach to say that you have to use this attribute.

I think we can all discuss how we can resolve this issue if this change is to be reverted. But my expectation is that this example, which works in PF11, PF10 and 8, also works in new versions without any attribute obligation. Perhaps the newly added changes should be tied to the option so that they do not affect the old usage.

+1 for reverting it if we are going to solve this problem in a different way otherwise -1.

Test project: https://github.com/primefaces/primefaces/issues/9349#issuecomment-1293407918

**Comment by mertsincan** [comment]
https://github.com/primefaces/primefaces/issues/9349#issuecomment-1304288509 I agree with @djmj

**Comment by melloware** [comment]
So it sounds like it should be a config and defaulted to legacy behavior?

**Comment by Rapster** [comment]
Ok, I understand after such a long exchanges, we can get lost... I could have remove the "noise" to make things more clear, but not really confortable doing that 😅 

I'll try to make things clear, concise and simple. Problems/Solutions are the following (based on what I said):

> - once one/several value(s) from the custom filter is/are selected, it/they gets lost in column filter

Solution is to use `oncomplete`. More details [here](https://github.com/primefaces/primefaces/issues/9349#issuecomment-1299789421)

> - if you have preselected values in the custom filter (e.g selectCheckboxMenu here), these are not showing in the filter therefore not taken account in load method at page loading

Solution is to use `Column#filterValue`, more details [here](https://github.com/primefaces/primefaces/issues/9349#issuecomment-1304304966)

However, I think there is room for improvement. First, maybe improve the doc on  `Column#filterValue` to mention it stores the filter value, therefore can be used to set the default value. Second, as an enhancement, to avoid code redundancy _(see sample I gave)_, when custom filter facet is provided, `FilterMeta#filterValue` could be initlialized with the facet value (`ValueHolder#getValue`) so users don't have to use `Column#filterValue`

How does it sound to you all?

**Comment by mertsincan** [comment]
I'm not sure, but I think if there really is going to be an attribute like filterValue, we should use it every time we use filter. An approach such as use this for these cases and use another attribute for other cases is utterly confusing. users may be undecided about what to use in which situation.

**Comment by Rapster** [comment]
I'm really trying to help you here... 😅 

>  we should use it every time we use filter

Only when you have default values, yes 

> An approach such as use this for these cases and use another attribute for other cases is utterly confusing.

Well on one hand you don't like the use of filterValue but on the other hand, you want it to be used every time. I'm confused...

**Comment by mertsincan** [comment]
> I'm really trying to help you here... 😅
- I know, thank you :)

> Well on one hand you don't like the use of filterValue but on the other hand, you want it to be used every time. I'm confused...
- No, if it is necessary to use filterValue as you suggested above, it can be a bit confusing for what situations users (including me) should use filterValue. Frankly, I didn't know that filterValue should be used until I wrote the above solution :D
It would be great if we can find a solution without using filterValue like in old usage. But if we are going to need an attribute like filterValue for such issues, we should specify it to users like breaking change.

**Comment by tandraschko** [comment]
The LazyDataModel API is built to be generic and we pass all filters via FilterMeta in the #load method
so the normal case is that you loop all FilterMeta and add all in your "where" statement, like we do in our JpaLazyDataModel
this is the way how LazyDataModel was designed and worked in the past and also works now

the case, that you bind one single filter and use it in a non-generic (and "binding"-)way, is the special way, not the way how it should be used
also this makes it currently not possible to reset the filterValue like mentioned above (https://github.com/primefaces/primefaces/issues/9040)

so IMO we have 2 opinions:
1) revert your commit and just add it in the migration guide
2) introduce a new attribute which activates the way your commit works (i dont like this way that much, its up to you)

**Comment by hink084** [comment]
Hey guys!  I'm a little late to this party as I'm just now upgrading from v11 to v12.  

In regards to when or when not to utilize the `filterValue` attribute...I was under the impression that the attribute was to be used if a filter was not explicitly defined in a filter facet via `<f:facet name="filter">`. Or in other words, the `filterValue` attribute was to be ignored if I did define a filter facet.  

After upgrading to v12, I also discovered my filter values were empty in the UI when attempting to restore them during `@PostConstruct`.  This only occurred when the filter facet was explicitly defined, and I was not utilizing the `filterValue` attribute. (Which I wasn't doing because of Issue #8992 in v11.)

**Broken in v12:**
```html
<p:column filterBy="#{employee.title}"
          headerText="Title"
          id="col_title">

    <f:facet name="filter">
        <p:inputText onkeyup="PF('wgt_test').filter();" value="#{testView.titleFilterValue}"/>
    </f:facet>

    #{employee.title}
</p:column>
``` 

In v12, if I utilize the `filterValue` attribute (regardless if I define a filter facet or not) the values are restored.

**Working in v12:**
```html
<p:column filterBy="#{employee.title}"
          filterValue="#{testView.titleFilterValue}"
          headerText="Title"
          id="col_title">

    <f:facet name="filter">
        <p:inputText onkeyup="PF('wgt_test').filter();"/>
    </f:facet>

    #{employee.title}
</p:column>
``` 

```html
<p:column filterBy="#{employee.title}"
          filterValue="#{testView.titleFilterValue}"
          headerText="Title"
          id="col_title">
    #{employee.title}
</p:column>
``` 

I've gone ahead and attached another very simple/basic reproducer without a `LazyDataModel` to demonstrate this issue.  I figure it doesn't hurt to have another test case.

Obviously, this issue was caused by de4fe4a to fix Issue #9040.  It appears to have been resolved/reverted in v13.0.0-SNAPSHOT via 5fa4cbb, but since this issue is still open, I am unsure if you all decided that this was the long-term solution or not.

**Reproducer:**
[column-filtervalue.zip](https://github.com/primefaces/primefaces/files/11314771/column-filtervalue.zip)

**Comment by Rapster** [comment]
@hink084 Thanks for joining the party

Indeed, that something I mentionned before:

> Second, as an enhancement, to avoid code redundancy (see sample I gave), when custom filter facet is provided, FilterMeta#filterValue could be initlialized with the facet value (ValueHolder#getValue) so users don't have to use Column#filterValue

IMO facet filter (or facet in general?) should be about custom rendering, the rest is doable on `UIColumn`, but I understand things can get confusing since people are used to set default value with `InputText#value` for instance, but there is also `UIColumn#filterValue` (which is convenient for non custom filter)

So, maybe the best way to go about it is to support both by simply doing a coalesce 😄 I can propose a fix, but maybe in another issue, your issue is pretty clear as to what bothers you
